tests/manual/media/FXMediaPlayer/src/fxmediaplayer/media/FXMedia.java

```diff
@@ -0,0 +1,141 @@
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package fxmediaplayer.media;
+
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.nio.file.Path;
+import java.nio.file.FileSystem;
+import java.nio.file.FileSystems;
+import java.util.List;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.io.IOException;
+import java.util.stream.Stream;
+import java.nio.file.Files;
+import java.util.function.Predicate;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.nio.file.NoSuchFileException;
+
+/**
+ * Helper class to generate playlist from embeded media files located under
+ * src/fxmediaplayer/media. This is useful to test "file", "jar" and "jrt" protocol.
+ * Copy supported media files under src/fxmediaplayer/media,
+ * compile "FXMediaPlayer", run it and go to "Play List" tab and you should see
+ * your embeded media files.
+ *
+ * File protcol:
+ * cd rt/tests/manual/media/FXMediaPlayer
+ * and run
+ *
+ * JAR protcol:
+ * cd rt/tests/manual/media/FXMediaPlayer
+ * ant
+ * java @../../../../build/run.args -jar dist/FXMediaPlayer.jar
+ *
+ * JRT protocol:
+ * cd rt/tests/manual/media/FXMediaPlayer
+ * ant
+ * [macOS/Linux] jlink --output dist/FXMediaPlayer -p ../../../../build/jmods:dist --add-modules javafx.base,javafx.controls,javafx.media,javafx.graphics,FXMediaPlayer --launcher FXMediaPlayer=FXMediaPlayer/fxmediaplayer.FXMediaPlayer
+ * [macOS/Linux] ./dist/FXMediaPlayer/bin/FXMediaPlayer
+ * [Windows] jlink --output dist/FXMediaPlayer -p ../../../../build/jmods;dist --add-modules javafx.base,javafx.controls,javafx.media,javafx.graphics,FXMediaPlayer --launcher FXMediaPlayer=FXMediaPlayer/fxmediaplayer.FXMediaPlayer
+ * [Windows] dist\FXMediaPlayer\bin\FXMediaPlayer.bat
+ */
+public class FXMedia {
+
+    private static List<String> SUPPORTED_EXT = new ArrayList<>();
+    private static boolean isJRT = true;
+
+    static {
+        SUPPORTED_EXT.add("mp3");
+        SUPPORTED_EXT.add("wav");
+        SUPPORTED_EXT.add("aif");
+        SUPPORTED_EXT.add("mp4");
+        SUPPORTED_EXT.add("m4a");
+        SUPPORTED_EXT.add("m4v");
+    }
+
+    public static List<String> getEmbededMediaFiles() {
+        List<String> sources = new ArrayList<>();
+
+        try {
+            Path path = null;
+            URI uri = null;
+            Stream<Path> walk = null;
+            try {
+                FileSystem fileSystem = FileSystems.getFileSystem(URI.create("jrt:/"));
+                path = fileSystem.getPath("modules", "FXMediaPlayer",
+                        "fxmediaplayer", "media");
+                walk = Files.walk(path, 1);
+            } catch (NoSuchFileException ex) {
+                isJRT = false;
+                uri = FXMedia.class.getResource("/fxmediaplayer/media").toURI();
+
+                if (uri.getScheme().equals("jar")) {
```

Comment:
is it possible for uri scheme to be null?
may be "jar".equals(uri.getScheme()) ?

Reply:
Related question: Can `uri` itself be null?

Reply:
so the logic seems to be either a "jar", or else it's a file, right?
is it possible to be something else (i.e. the resource is missing altogether)? or some other condition?

Reply:
Yes, when we run jar file it will be jar and if we run unpacked it will be file. If resource is missing we will not enumerate files and it is acceptable condition. We should fallback to default playlist if no files present.